ext/enchant/enchant.c

```diff
@@ -138,65 +162,71 @@ static void php_enchant_list_dicts_fn( const char * const lang_tag,
 }
 /* }}} */
 
-static void php_enchant_broker_free(zend_resource *rsrc) /* {{{ */
-{
-	if (rsrc->ptr) {
-		enchant_broker *broker = (enchant_broker *)rsrc->ptr;
-		if (broker) {
-			if (broker->pbroker) {
-				if (broker->dictcnt && broker->dict) {
-					if (broker->dict) {
-						int total;
-						total = broker->dictcnt-1;
-						do {
-							if (broker->dict[total]) {
-								enchant_dict *pdict = broker->dict[total];
-								broker->dict[total] = NULL;
-								zend_list_free(pdict->rsrc);
-								efree(pdict);
-							}
-							total--;
-						} while (total>=0);
-					}
-					efree(broker->dict);
-					broker->dict = NULL;
-				}
-				enchant_broker_free(broker->pbroker);
-			}
-			efree(broker);
+static void php_enchant_broker_free(zend_object *object) /* {{{ */
+{
+	enchant_broker *broker = enchant_broker_from_obj(object);
+
+	if (broker) {
+		if (broker->pbroker) {  /* may have been freed by enchant_broker_free */
+			enchant_broker_free(broker->pbroker);
```

Comment:
Are you sure that it is safe to remove the code for freeing dictionaries first? The potential problem I see is that objects could be destroyed in the wrong order during GC or shutdown.

Reply:
Yes, as we have a copy in each dict (zbroker), which protect broker to be freed until all dict are freed. IIRC mysqli use the same solution (link_handle in each stmt)

Reply:
My concern here is specifically the GC case. Could we end up with the broker as a GC root? Maybe it's not possible.

Probably a way to ensure this is to forbid creation of dynamic properties on the class. We probably want to do that for all new internal classes we add, but especially those for resource -> object conversions.

Reply:
+1 (would be nice to be able to do that in userland as well, but https://wiki.php.net/rfc/locked-classes turned out to be too controversial)

Reply:
Sorry, but I don't see how this can be possible...  and even if the broker is destroyed before the dictionaries, the library will properly free the open dicts.